pkg/alertmanager/distributor.go

```diff
@@ -0,0 +1,272 @@
+package alertmanager
+
+import (
+	"context"
+	"hash/fnv"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"sync"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/weaveworks/common/httpgrpc"
+	"github.com/weaveworks/common/httpgrpc/server"
+	"github.com/weaveworks/common/user"
+
+	"github.com/cortexproject/cortex/pkg/alertmanager/alertmanagerpb"
+	"github.com/cortexproject/cortex/pkg/ring"
+	"github.com/cortexproject/cortex/pkg/ring/client"
+	"github.com/cortexproject/cortex/pkg/tenant"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/services"
+)
+
+// Distributor forwards requests to individual alertmanagers.
+type Distributor struct {
+	services.Service
+
+	cfg              *MultitenantAlertmanagerConfig
+	requestsInFlight sync.WaitGroup
+
+	alertmanagerRing        ring.ReadRing
+	alertmanagerClientsPool ClientsPool
+
+	logger log.Logger
+
+	receivedRequests *prometheus.CounterVec
+	amSends          *prometheus.CounterVec
+	amSendFailures   *prometheus.CounterVec
+}
+
+// NewDistributor constructs a new Distributor
+func NewDistributor(cfg *MultitenantAlertmanagerConfig, alertmanagersRing *ring.Ring, alertmanagerClientsPool ClientsPool, logger log.Logger, reg prometheus.Registerer) (d *Distributor, err error) {
+	if alertmanagerClientsPool == nil {
+		alertmanagerClientsPool = newAlertmanagerClientsPool(client.NewRingServiceDiscovery(alertmanagersRing), cfg.AlertmanagerClient, logger, reg)
+	}
+
+	d = &Distributor{
+		cfg:                     cfg,
+		logger:                  logger,
+		alertmanagerRing:        alertmanagersRing,
+		alertmanagerClientsPool: alertmanagerClientsPool,
+	}
+
+	d.receivedRequests = promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+		Name: "cortex_alertmanager_distributor_received_requests_total",
+		Help: "The total number of requests received.",
+	}, []string{"user"})
+	d.amSends = promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+		Name: "cortex_alertmanager_distributor_alertmanager_send_total",
+		Help: "The total number of requests sent to the alertmanager.",
+	}, []string{"ingester"})
+	d.amSendFailures = promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+		Name: "cortex_alertmanager_distributor_alertmanager_send_failures_total",
+		Help: "The total number of requests sent to the alertmanager that failed.",
+	}, []string{"ingester"})
+
+	d.Service = services.NewBasicService(d.starting, d.running, d.stopping)
```

Comment:
Since `starting()` and `stopping()` are empty functions, you can remove them and pass `nil` here. It's supported by the service implementation.